consensus/cometbft/service/finalize_block.go

```diff
@@ -48,52 +54,91 @@ func (s *Service) finalizeBlockInternal(
-	} else {
-		// Preserve the CosmosSDK context while using the correct base ctx.
-		s.finalizeBlockState.SetContext(s.finalizeBlockState.Context().WithContext(ctx))
+	// Check whether currently block hash is already available. If so
+	// we may speed up block finalization.
+	hash := string(req.Hash)
+	switch cached, err := s.cachedStates.GetCached(hash); {
+	case err == nil:
+		// Block with height equal to initial height is special and we can't rely on its cache.
+		// This is because Genesis state is cached but not committed (and purged from s.cachedStates)
+		// We handle the case below.
+		if req.Height > s.initialHeight {
+			if err = s.cachedStates.MarkAsFinal(hash); err != nil {
+				return nil, fmt.Errorf("failed marking state as final, hash %s, height %d: %w", hash, req.Height, err)
+			}
+
+			finalState := cached.State
+			var (
+				signedBlk *ctypes.SignedBeaconBlock
+				sidecars  datypes.BlobSidecars
+			)
+			signedBlk, sidecars, err = s.Blockchain.ParseBeaconBlock(req)
+			if err != nil {
+				return nil, fmt.Errorf("finalize block: failed parsing block: %w", err)
+			}
+			blk := signedBlk.GetBeaconBlock()
+
+			if err = s.Blockchain.FinalizeSidecars(
+				finalState.Context(),
+				req.SyncingToHeight,
+				blk,
+				sidecars,
+			); err != nil {
+				return nil, fmt.Errorf("failed finalizing sidecars: %w", err)
+			}
+			if err = s.Blockchain.PostFinalizeBlockOps(
+				finalState.Context(),
+				blk,
+			); err != nil {
+				return nil, fmt.Errorf("finalize block: failed post finalize block ops: %w", err)
+			}
+			return formatFinalizeBlockResponse(cached.ValUpdates, len(req.Txs), s.cmtConsensusParams)
+		}
+
+	case errors.Is(err, cache.ErrStateNotFound):
+		// this is a benign error, it just signal it's the first time we see the block being finalized
+		// Keep processing below
+
+	default:
+		return nil, fmt.Errorf("failed checking cached state, hash %s, height %d: %w", hash, req.Height, err)
 	}
 
-	// This result format is expected by Comet. That actual execution will happen as part of the state transition.
-	txResults := make([]*cmtabci.ExecTxResult, len(req.Txs))
-	for i := range req.Txs {
-		//nolint:mnd // its okay for now.
-		txResults[i] = &cmtabci.ExecTxResult{
-			Codespace: "sdk",
-			Code:      2,
-			Log:       "skip decoding",
-			GasWanted: 0,
-			GasUsed:   0,
-		}
+	// Block has not been cached already, as it happens when we block sync.
+	// Normally we would directly process it but first block post initial height
+	// needs special handling.
+	var finalState *cache.State
+	cachedFinalHash, cachedFinalState, err := s.cachedStates.GetFinal()
+	switch {
+	case errors.Is(err, cache.ErrNoFinalState):
+		hash = string(req.Hash) // not necessary but clarifies intent
+		finalState = s.resetState(ctx)
+
+	case err == nil:
+		hash = cachedFinalHash
+		finalState = cachedFinalState
+
+		// Preserve the CosmosSDK context while using the correct base ctx.
+		finalState.SetContext(finalState.Context().WithContext(ctx))
```

Comment:
this is the special case when we finalize the first block. To simplify implementation I chose not to properly cache it but to execute it all in FinalizeBlock, similarly to what we do today.